### PR TITLE
refactor: oAuth proxy through HTTP kernel

### PR DESCRIPTION
**Pull request by alirezamirsepassi**
First of all thanks so much for taking the time to open a pull request and help the project. It's because of people like you that we love working on this project.

Please read the list below. Feel free to delete this text after but we need you to read it so we make sure that the project is consistent and remains of quality.

#### Before submitting the PR
- [x] Read the [CONTRIBUTING document](https://github.com/monicahq/monica/blob/master/CONTRIBUTING.md) before submitting your PR.
- [x] If the PR is related to an issue or fix one, don't forget to indicate it.
- [x] Indicate `[wip]` in the title of the PR it is is not final yet. Remove `[wip]` when ready. Otherwise the PR will be considered complete and rejected if it's not working.

### General checks
- [x] Make sure that the change you propose is the smallest possible.
- [x] The name of the PR should follow the [conventional commits guideline](https://github.com/monicahq/monica/blob/master/docs/contribute/index.md#conventional-commits) that the project follows.

### Front-end changes
- [x] If you change the UI, make sure to ask repositories administrators first about your changes by pinging @djaiss or @asbiin in this PR.
- [x] Screenshots are included if the PR changes the UI.
- [x] Front-end tests have been written with Cypress.

#### Backend/models changes
- [x] The API has been updated.
- [x] API's documentation has been added by submitting a pull request in the [marketing website repository](https://github.com/monicahq/marketing_site/pulls).
- [x] Tests have been added for the new code.
- [x] If you change a model, make sure the FakeContentTableSeeder is updated. We need seeders to develop locally and generate fake data.

#### If the code changes the SQL schema
- [x] Make sure exporting account data as SQL is still working.
- [x] Make sure your changes do not break importing data with `vCard` and `.csv` files.
- [x] Make sure account reset and deletion still work.

#### Other tasks
- [x] [CHANGELOG](https://github.com/monicahq/monica/blob/master/CHANGELOG.md) entry added, if necessary, under `UNRELEASED`.
- [x] [CONTRIBUTORS](https://github.com/monicahq/monica/blob/master/CONTRIBUTORS) entry added, if necessary.
- [x] If it's relevant and worth mentioning, create a changelog entry for this change. The changelog entry will appear inside the UI for all users to see. To know if your change is worth the creation of a changelog entry, [read the documentation](https://github.com/monicahq/monica/blob/master/docs/administrators/tips.md#when-is-it-relevant-to-create-a-changelog-entry).
- [x] Don't forget to [ask for a free account](mailto:regis@monicahq.com) on https://monicahq.com as anyone who contributes can request a free account.

### CODE REVIEW

**Comment by alirezamirsepassi** [comment]
This way your application can work with a single thread, also a more convenient way to do subrequests.

File: app/Http/Controllers/Api/Auth/OAuthController.php [hunk 1]
```diff
@@ -129,26 +128,29 @@ private function validateRequest(Request $request)
...
-                'password' => $data['password'],
-                'scope' => '',
-            ],
-        ]);
-
-        $data = json_decode($response->getBody());
+        $http = app(\Illuminate\Contracts\Http\Kernel::class);
+        $response = $http->handle(
+            \Illuminate\Http\Request::create(
+                '/oauth/token',
```
Comment (asbiin) [inline app/Http/Controllers/Api/Auth/OAuthController.php]: Could you use the `$url` var define before? I prefer to use `route` when it's possible, and here I'm not sure if it's working with subdirectory installations.

Comment (alirezamirsepassi) [reply app/Http/Controllers/Api/Auth/OAuthController.php]: Sorry, I can't understand what is wrong with the style ci. Can you help me on it ?

File: app/Http/Controllers/Api/Auth/OAuthController.php [hunk 2]
```diff
@@ -68,6 +67,30 @@ public function login(Request $request)
...
 
+    /**
+     * Validate the request.
+     *
+     * @param Request $request
+     * @return \Illuminate\Http\JsonResponse|true
+     */
+    private function validateRequest(Request $request)
+    {
+        $validator = Validator::make($request->all(), ['email' => 'email|required', 'password' => 'required',]);
```
Comment (asbiin) [inline app/Http/Controllers/Api/Auth/OAuthController.php]: Please indent like this:
```suggestion
        $validator = Validator::make($request->all(), [
            'email' => 'email|required',
            'password' => 'required',
        ]);
```

File: app/Http/Controllers/Api/Auth/OAuthController.php [hunk 3]
```diff
@@ -68,6 +67,30 @@ public function login(Request $request)
...
+
+        if ($validator->fails()) {
+            return $this->respondValidatorFailed($validator);
+        }
+
+        // Check if email exists. If not respond with an Unauthorized, this way a hacker
+        // doesn't know if the login email exist or not, or if the password os wrong
```
Comment (asbiin) [inline app/Http/Controllers/Api/Auth/OAuthController.php]: One typo
```suggestion
        // doesn't know if the login email exist or not, or if the password is wrong
```

File: app/Http/Controllers/Api/Auth/OAuthController.php [hunk 4]
```diff
@@ -85,74 +108,37 @@ public function verify(Request $request)
         }
 
         try {
-            $token = $this->proxy([
-                'username' => $request->input('email'),
-                'password' => $request->input('password'),
-                'grantType' => 'password',
-            ]);
+            $token = $this->proxy(['username' => $request->input('email'), 'password' => $request->input('password'), 'grantType' => 'password',]);
```
Comment (asbiin) [inline app/Http/Controllers/Api/Auth/OAuthController.php]: Don't remove indentation
```suggestion
            $token = $this->proxy([
                'username' => $request->input('email'),
                'password' => $request->input('password'),
                'grantType' => 'password',
            ]);
```

File: app/Http/Controllers/Api/Auth/OAuthController.php [hunk 5]
```diff
@@ -85,74 +108,37 @@ public function verify(Request $request)
...
-
-        $data = json_decode($response->getBody());
-
-        return [
-            'access_token' => $data->access_token,
-            'expires_in' => $data->expires_in,
-        ];
+        $url = App::runningUnitTests() ? config('app.url') . '/oauth/token' : route('passport.token');
+        $http = app(Kernel::class);
+        $response = $http->handle(Request::create($url, 'POST', [
```
Comment (asbiin) [inline app/Http/Controllers/Api/Auth/OAuthController.php]: You could remove intermidiate `http` variable:
```suggestion
        $response = app(Kernel::class)->handle(Request::create($url, 'POST', [
```

File: app/Http/Controllers/Api/Auth/OAuthController.php [hunk 6]
```diff
@@ -85,74 +108,37 @@ public function verify(Request $request)
...
+            'client_id' => config('monica.mobile_client_id'),
+            'client_secret' => config('monica.mobile_client_secret'),
+            'username' => $data['username'],
+            'password' => $data['password'],
+            'scope' => ''
+        ]));
+
+        $data = json_decode($response->content());
+
+        return ['access_token' => $data->access_token, 'expires_in' => $data->expires_in,];
```
Comment (asbiin) [inline app/Http/Controllers/Api/Auth/OAuthController.php]: Please don't remove indentation:
```suggestion
        return [
            'access_token' => $data->access_token,
            'expires_in' => $data->expires_in,
        ];
```

File: app/Http/Controllers/Api/Auth/OAuthController.php [hunk 7]
```diff
@@ -2,16 +2,15 @@
 
 namespace App\Http\Controllers\Api\Auth;
 
-use GuzzleHttp\Client;
+use App\Http\Controllers\Controller;
 use App\Models\User\User;
+use App\Traits\JsonRespondController;
+use Barryvdh\Debugbar\Facade as Debugbar;
+use Illuminate\Contracts\Http\Kernel;
```
Comment (asbiin) [inline app/Http/Controllers/Api/Auth/OAuthController.php]: You have to add the removed line:
```
use function Safe\json_decode;
```

**Comment by asbiin** [comment]
Great!
After fixing the latest suggestions, you may push the commit to let styleCI run again. Then you can download a raw diff from StyleCI here:
image

and apply this diff with: `patch -p1 < file.patch`

**Comment by alirezamirsepassi** [comment]
All done. Thank you for helping.

**Comment by asbiin** [comment]
@alirezamirsepassi  Thank you for your help!

**Pull request merged**